Fix #2992 (restore mechanism on docker volumes)

hgross opened the pull request:
<!--
Thank you for your interest into contributing to OctoPrint, it's
highly appreciated!

Please make sure you have read the "guidelines for contributing" as
linked just above this form, there's a section on Pull Requests in there
as well that contains important information.

As a summary, please make sure you have ticked all points on this
checklist:
-->

  * [x] Your changes are not possible to do through a plugin and relevant
    to a large audience (ideally all users of OctoPrint)
  * [x] If your changes are large or otherwise disruptive: You have
    made sure your changes don't interfere with current development by
    talking it through with the maintainers, e.g. through a
    Brainstorming ticket
  * [x] Your PR targets OctoPrint's devel branch, or maintenance if it's
    a bug fix for an issue present in the current stable version (no PRs
    against master or anything else please)
  * [x] Your PR was opened from a custom branch on your repository
    (no PRs from your version of master, maintenance or devel please),
    e.g. dev/my_new_feature or fix/my_bugfix
  * [x] Your PR only contains relevant changes: no unrelated files,
    no dead code, ideally only one commit - rebase and squash your PR 
    if necessary!
  * [x] Your changes follow the existing coding style
  * [x] If your changes include style sheets: You have modified the
    .less source files, not the .css files (those are generated with
    lessc)
  * [x] You have tested your changes (please state how!) - ideally you
    have added unit tests
  * [x] You have run the existing unit tests against your changes and
    nothing broke
  * [x] You have added yourself to the AUTHORS.md file :)

<!--
Describe your PR further using the template provided below. The more 
details the better!
-->

#### What does this PR do and why is it necessary?
fixes #2992 

#### How was it tested? How can it be tested by the reviewer?
Steps to reproduce are stated in #2992

#### Any background context you want to provide?
More on docker-volumes: https://docs.docker.com/storage/volumes/#choose-the--v-or---mount-flag
I am using this with resin.io/balena-cloud

#### What are the relevant tickets if any?
#2992

#### Screenshots (if appropriate)
see logs in #2992

#### Further notes
-

tedder commented:
Yeah, as you found, `os.rename` doesn't work because that changes hard links, and hard links must be on the same volume.

What are you using for a base Dockerfile?

hgross commented:
Switched PR to the maintenance branch.
I use [balena.io](https://www.balena.io/) for deployment - this is the Dockerfile (part of a microservice deployment):

```
FROM resin/raspberrypi3-debian:jessie

RUN apt-get update && apt-get install -y --no-install-recommends \
	python-pip \
	python-dev \
	git \
	build-essential \
	libav-tools \
	avrdude \
	curl \
	subversion \
	libjpeg8-dev \
	imagemagick \
	libav-tools \
	cmake \
  && apt-get clean \
  && rm -rf /tmp/* /var/tmp/*  \
  && rm -rf /var/lib/apt/lists/*

# Move to app dir
WORKDIR /usr/src/app

# Install Octoprint from source
RUN git clone https://github.com/hgross/OctoPrint ./octoprint \
  && cd /usr/src/app/octoprint && pip install --upgrade pip && pip install setuptools \
  && cd /usr/src/app/octoprint && pip install -r requirements.txt && python setup.py install \
  && cd /usr/src/app/octoprint && pip install https://github.com/BillyBlaze/OctoPrint-TouchUI/archive/master.zip \
	&& cd /usr/src/app/octoprint && pip install https://github.com/marian42/octoprint-preheat/archive/master.zip \
	&& cd /usr/src/app/octoprint && pip install "https://github.com/OctoPrint/OctoPrint-DisplayProgress/archive/master.zip" \
	&& cd /usr/src/app/octoprint && pip install https://github.com/malnvenshorn/OctoPrint-CostEstimation/archive/master.zip \
	&& cd /usr/src/app/octoprint && pip install "https://github.com/imrahil/OctoPrint-PrintHistory/archive/master.zip" \
	&& cd /usr/src/app/octoprint && pip install "https://github.com/ziirish/OctoPrint-TuyaSmartplug/archive/master.zip" \
	&& cd /usr/src/app/octoprint && pip install "https://github.com/eyal0/OctoPrint-PrintTimeGenius/archive/master.zip" \
	&& cd /usr/src/app/octoprint && pip install "https://github.com/malnvenshorn/OctoPrint-FilamentManager/archive/master.zip"


	


# Move app to filesystem
COPY ./app ./

## uncomment if you want systemd
ENV INITSYSTEM on

# Start app
CMD ["bash", "/usr/src/app/start.sh"]

```

When my setup is done I will clean that up and share it - including the pi-camera stream and fixed plugin/octoprint versions.

foosel commented:
Odd... When I just looked at that, I was sure I had already done this. For some reason I didn't. Ah well, thanks for the PR! :)
